contracts/zero-ex/contracts/src/features/TransformERC20.sol

```diff
@@ -0,0 +1,371 @@
+            callDataHash,
+            taker,
+            inputToken,
+            outputToken,
+            inputTokenAmount,
+            minOutputTokenAmount,
+            transformations
+        );
+    }
+
+    /// @dev Private version of `transformERC20()`.
+    /// @param callDataHash Hash of the ingress calldata.
+    /// @param taker The taker address.
+    /// @param inputToken The token being provided by the taker.
+    ///        If `0xeee...`, ETH is implied and should be provided with the call.`
+    /// @param outputToken The token to be acquired by the taker.
+    ///        `0xeee...` implies ETH.
+    /// @param inputTokenAmount The amount of `inputToken` to take from the taker.
+    ///        If set to `uint256(-1)`, the entire spendable balance of the taker
+    ///        will be solt.
+    /// @param minOutputTokenAmount The minimum amount of `outputToken` the taker
+    ///        must receive for the entire transformation to succeed. If set to zero,
+    ///        the minimum output token transfer will not be asserted.
+    /// @param transformations The transformations to execute on the token balance(s)
+    ///        in sequence.
+    /// @return outputTokenAmount The amount of `outputToken` received by the taker.
+    function _transformERC20Private(
+        bytes32 callDataHash,
+        address payable taker,
+        IERC20TokenV06 inputToken,
+        IERC20TokenV06 outputToken,
+        uint256 inputTokenAmount,
+        uint256 minOutputTokenAmount,
+        Transformation[] memory transformations
+    )
+        public
+        payable
+        returns (uint256 outputTokenAmount)
+    {
+        // If the input token amount is -1, transform the taker's entire
+        // spendable balance.
+        if (inputTokenAmount == uint256(-1)) {
+            inputTokenAmount = ITokenSpender(address(this))
+                .getSpendableERC20BalanceOf(inputToken, taker);
+        }
+
+        IFlashWallet wallet = getTransformWallet();
+
+        // Remember the initial output token balance of the taker.
+        uint256 takerOutputTokenBalanceBefore =
+            LibERC20Transformer.getTokenBalanceOf(outputToken, taker);
+
+        // Pull input tokens from the taker to the wallet and transfer attached ETH.
+        _transferInputTokensAndAttachedEth(inputToken, taker, address(wallet), inputTokenAmount);
+
+        // Perform transformations.
+        for (uint256 i = 0; i < transformations.length; ++i) {
+            _executeTransformation(wallet, transformations[i], taker, callDataHash);
+        }
+
+        // Compute how much output token has been transferred to the taker.
+        uint256 takerOutputTokenBalanceAfter =
+            LibERC20Transformer.getTokenBalanceOf(outputToken, taker);
+        if (takerOutputTokenBalanceAfter > takerOutputTokenBalanceBefore) {
+            outputTokenAmount = takerOutputTokenBalanceAfter.safeSub(
+                takerOutputTokenBalanceBefore
+            );
+        }
```

Comment:
I wonder if we should revert if the final balance is less than the initial balance? What's a scenario where the user would want this?

Reply:
Yeah, I can't really think of a compelling scenario.

Reply:
We now revert.